openedx/core/djangoapps/user_api/preferences/tests/test_api.py

```diff
@@ -0,0 +1,324 @@
+        """
+        Verifies that set_user_preferences returns appropriate errors.
+        """
+        update_data = {
+            self.test_preference_key: "new_value"
+        }
+        with self.assertRaises(UserNotFound):
+            update_user_preferences(self.user, update_data, username="no_such_user")
+
+        with self.assertRaises(UserNotFound):
+            update_user_preferences(self.no_such_user, update_data)
+
+        with self.assertRaises(UserNotAuthorized):
+            update_user_preferences(self.staff_user, update_data, username=self.user.username)
+
+        with self.assertRaises(UserNotAuthorized):
+            update_user_preferences(self.different_user, update_data, username=self.user.username)
+
+        too_long_key = "x" * 256
+        with self.assertRaises(PreferenceValidationError):
+            update_user_preferences(self.user, { too_long_key: "new_value"})
+
+        user_preference_save.side_effect = [Exception, None]
+        with self.assertRaises(PreferenceUpdateError):
+            update_user_preferences(self.user, { self.test_preference_key: "new_value"})
+
+        user_preference_delete.side_effect = [Exception, None]
+        with self.assertRaises(PreferenceUpdateError):
+            update_user_preferences(self.user, { self.test_preference_key: None })
+
+    def test_delete_user_preference(self):
+        """
+        Verifies the basic behavior of delete_user_preference.
+        """
+        self.assertTrue(delete_user_preference(self.user, self.test_preference_key))
+        self.assertFalse(delete_user_preference(self.user, "no_such_key"))
+
+    @patch('openedx.core.djangoapps.user_api.models.UserPreference.delete')
+    def test_delete_user_preference_errors(self, user_preference_delete):
+        """
+        Verifies that delete_user_preference returns appropriate errors.
+        """
+        with self.assertRaises(UserNotFound):
+            delete_user_preference(self.user, self.test_preference_key, username="no_such_user")
+
+        with self.assertRaises(UserNotFound):
+            delete_user_preference(self.no_such_user, self.test_preference_key)
+
+        with self.assertRaises(UserNotAuthorized):
+            delete_user_preference(self.staff_user, self.test_preference_key, username=self.user.username)
+
+        with self.assertRaises(UserNotAuthorized):
+            delete_user_preference(self.different_user, self.test_preference_key, username=self.user.username)
+
+        user_preference_delete.side_effect = [Exception, None]
+        with self.assertRaises(PreferenceUpdateError):
+            delete_user_preference(self.user, self.test_preference_key)
+
+
+@ddt.ddt
+class UpdateEmailOptInTests(ModuleStoreTestCase):
+
+    USERNAME = u'frank-underwood'
+    PASSWORD = u'ṕáśśẃőŕd'
+    EMAIL = u'frank+underwood@example.com'
+
+    def test_update_and_retrieve_preference_info(self):
+        # TODO: move test into preferences API test.
```

Comment:
@andy-armstrong see this TODO-- this was an old test point, but don't think there is any value in keeping it. Just delete?